File: python/c19_1_eulerian_paths.py
import unittest
from collections import defaultdict
from util import reversed_adj
import sys
from itertools import pairwise

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def existence(adj):
    odd=[]
    for i,e in enumerate(adj):
        if len(e)%2==1:
            odd.append(i)
    if len(odd)==0:
        return [-1,-1]
    elif len(odd)==2:
        return odd
    else:
        return None

def existence2(adj,rev=None):
    if rev is None:
        rev=reversed_adj(adj)
    d=defaultdict(list)
    for e,(i,o) in enumerate(zip(rev,adj)):
        d[len(i)-len(o)].append(e)
    if len(d[0])==len(adj):
        return (-1,-1)
    elif len(d[0])==len(adj)-2 and len(d[1])==1:
        return (d[-1][0],d[1][0])
    else:
        return None

def hierhorzer(adj):
    e=existence(adj)
    if e is None or not adj:
        return ()
    adj=[*map(set,adj)]
    def dfs(s):
        while adj[s]:
            u=next(iter(adj[s]))
            adj[s].remove(u)
            adj[u].remove(s)
            yield from dfs(u)
        yield s
    return dfs(max(0,e[1]))

def hierhorzer2(adj):
    rev=reversed_adj(adj)
    e=existence2(adj,rev)
    if not e or not adj:
        return ()
    def dfs(s):
        while rev[s]:
            yield from dfs(rev[s].pop())
        yield s
    return dfs(max(0,e[1]))


class Test(unittest.TestCase):
    def assertHierhorzer(self,adj):
        with self.subTest(adj=adj):
            adj2=[[] for _ in adj]
            for i,j in pairwise(hierhorzer(adj)):
                adj2[i].append(j)
                adj2[j].append(i)
            for i,j in zip(adj,adj2):
                self.assertEqual(sorted(i), sorted(j))
    def assertHierhorzer2(self,adj):
        with self.subTest(adj=adj):
            adj2=[[] for _ in adj]
            for i,j in pairwise(hierhorzer2(adj)):
                adj2[i].append(j)
            for i,j in zip(adj,adj2):
                self.assertEqual(sorted(i), sorted(j))
    def test(self):
        adj1=[[1,3],[0,2,4],[1,4],[0,4],[1,2,3]]
        adj2=[[1,3],[0,2,3,4],[1,4],[0,1],[1,2]]

        self.assertEqual(existence([]),[-1,-1])
        self.assertCountEqual(existence(adj1),[1,4])
        self.assertEqual(existence(adj2),[-1,-1])
        self.assertIsNone(existence([[1,2,3],[0,2,3],[0,1,3],[0,1,2]]))

        adj3=[[1],[2,4],[4],[0],[3]]
        adj4=[[3],[0,4],[1],[1],[2]]

        self.assertEqual(existence2([]),(-1,-1))
        self.assertEqual(existence2(adj3),(1,4))
        self.assertEqual(existence2(adj4),(-1,-1))
        self.assertIsNone(existence2([[1,2,3],[2,3],[3],[]]))

        self.assertHierhorzer([])
        self.assertHierhorzer(adj1)
        self.assertHierhorzer(adj2)
        self.assertHierhorzer([[1,2],[0,2,4,5],[0,1,3,5],[2,6],[1,5],[1,2,4,6],[3,5]])

        self.assertHierhorzer2([])
        self.assertHierhorzer2(adj3)
        self.assertHierhorzer2(adj4)
        self.assertHierhorzer2([[1],[2,4],[0,3],[6],[5],[1,2],[5]])

if __name__=='__main__':
    unittest.main()

File: c++/c19_1_eulerian_paths.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c19_1_eulerian_paths.h"
#include "util.h"

namespace {
	optional<pair<ll, ll>> existence(const vector<vector<ll>>& adj) {
		vector<ll> odd;
		for (ll i = 0; i < ssize(adj); ++i) {
			if (adj[i].size() % 2 == 1) {
				odd.push_back(i);
			}
		}
		if (odd.size() == 2)
			return make_pair(odd[0], odd[1]);
		else if (odd.size() == 0)
			return make_pair(-1, -1);
		else
			return nullopt;
	}

	optional<pair<ll, ll>> existence2(const vector<vector<ll>>& adj) {
		auto rev = reversed_adj(adj);
		unordered_map<ll, vector<ll>> d;
		for (ll i = 0; i < ssize(adj); ++i)
			d[ssize(rev[i]) - ssize(adj[i])].push_back(i);
		if (d[0].size() == adj.size() - 2 && d.contains(1) && d.contains(-1))
			return make_pair(d[-1][0], d[1][0]);
		else if (d[0].size() == adj.size())
			return make_pair(-1, -1);
		else
			return nullopt;
	}

	vector<ll> hierholzer(const vector<vector<ll>>& adj) {
		auto e = existence(adj);
		if (!e || adj.empty())
			return {};
		auto adj2 = adj | views::transform(ranges::to<unordered_set>()) | ranges::to<vector>();
		vector<ll> ret;
		auto dfs = [&](this auto self, ll s) -> void {
			while (!adj2[s].empty()) {
				ll u = *adj2[s].begin();
				adj2[s].erase(u);
				adj2[u].erase(s);
				self(u);
			}
			ret.push_back(s);
			};
		dfs(max(0ll, e->second));
		return ret;
	}

	void test_hierholzer(const vector<vector<ll>>& adj) {
		vector<vector<ll>> adj2(adj.size());
		for (auto [i, j] : hierholzer(adj) | views::pairwise) {
			adj2[i].push_back(j);
			adj2[j].push_back(i);
		}
		for (ll i = 0; i < ssize(adj); ++i)
			EXPECT_EQ(adj[i] | ranges::to<unordered_multiset>(), adj2[i] | ranges::to<unordered_multiset>());
	}

	void test_hierholzer2(const vector<vector<ll>>& adj) {
		vector<vector<ll>> adj2(adj.size());
		for (auto [i, j] : hierholzer2(adj) | views::pairwise) {
			adj2[i].push_back(j);
		}
		for (ll i = 0; i < ssize(adj); ++i)
			EXPECT_EQ(adj[i] | ranges::to<unordered_multiset>(), adj2[i] | ranges::to<unordered_multiset>());
	}

	optional<pair<ll, ll>> convert(optional<pair<ll, ll>> p) {
		if (!p || p->first <= p->second)
			return p;
		else
			return make_pair(p->second, p->first);
	}
}

vector<ll> hierholzer2(vector<vector<ll>> adj) {
	auto e = existence2(adj);
	if (!e || adj.empty())
		return {};
	vector<ll> ret;
	auto dfs = [&](this auto self, ll s) -> void {
		while (!adj[s].empty()) {
			ll u = adj[s].back();
			adj[s].pop_back();
			self(u);
		}
		ret.push_back(s);
		};
	dfs(max(0ll, e->first));
	ranges::reverse(ret);
	return ret;
}

TEST(C191EulerianPaths, existence) {
	EXPECT_EQ(convert(existence({})), make_pair(-1, -1));
	EXPECT_EQ(convert(existence({ {1,3},{0,2,4},{1,4},{0,4},{1,2,3} })), make_pair(1, 4));
	EXPECT_EQ(convert(existence({ {1,3},{0,2,3,4},{1,4},{0,1},{1,2} })), make_pair(-1, -1));
	EXPECT_EQ(convert(existence({ {1,2,3},{0,2,3},{0,1,3},{0,1,2} })), nullopt);
}

TEST(C191EulerianPaths, existence2) {
	EXPECT_EQ(existence2({}), make_pair(-1, -1));
	EXPECT_EQ(existence2({ {1},{2,4},{4},{0},{3} }), make_pair(1, 4));
	EXPECT_EQ(existence2({ {3},{0,4},{1},{1},{2} }), make_pair(-1, -1));
	EXPECT_EQ(existence2({ {1,2,3},{2,3},{3},{} }), nullopt);
}

TEST(C191EulerianPaths, hierholzer) {
	test_hierholzer({});
	test_hierholzer({ {1,3},{0,2,4},{1,4},{0,4},{1,2,3} });
	test_hierholzer({ {1,3},{0,2,3,4},{1,4},{0,1},{1,2} });
	test_hierholzer({ {1,2},{0,2,4,5},{0,1,3,5},{2,6},{1,5},{1,2,4,6},{3,5} });
}

TEST(C191EulerianPaths, hierholzer2) {
	test_hierholzer2({});
	test_hierholzer2({ {1},{2,4},{4},{0},{3} });
	test_hierholzer2({ {3},{0,4},{1},{1},{2} });
	test_hierholzer2({ {1},{2,4},{0,3},{6},{5},{1,2},{5} });
}
